### PR TITLE
Fix handling multi valued fields

### PR DESCRIPTION
**Pull request by snejus**
While working on #6367 I noticed that users are currently required to use our internal separator `\␀` in order to edit multi-valued fields, for example `beet modify artists='a\␀b'`.

Similarly, this separator is used in output, for example, reporting of field changes:

```
$ beet modify path::aaa artists='a\␀b'
Modifying 8 items.
54898 | 2022 / RAVE SLUTZ: Fallen Shrine & dj Christian NXC - DEEEJAAAY
  artists: Fallen Shrine\␀dj Christian NXC -> a\␀b
```

This PR replaces `\␀` separator with `; ` for input and formats changes in multi-valued fields clearly:

```
$ beet modify path::aaa artists='a; b'
54898 | 2022 / RAVE SLUTZ: Fallen Shrine & dj Christian NXC - DEEEJAAAY
  artists:
    - Fallen Shrine
    - dj Christian NXC
    + a
    + b
```
image

### Architecture-level changes

- `DelimitedString` now separates concerns between:
  - database serialization via `db_delimiter` (`to_sql`)
  - user-facing/template formatting via a fixed `'; '` delimiter (`format`)
  - parsing that accepts both DB and user-facing separators (`parse`)
- Field diff rendering now has a dedicated path for list fields:
  - `_field_diff` detects list values
  - `_multi_value_diff` computes set-based added/removed entries and renders per-item diff lines
- Coloring responsibilities were streamlined:
  - raw ANSI application moved to `_colorize`
  - `colorize` is now only the feature-flag/environment gate
  - `colordiff` is reduced to string diff highlighting logic, with redundant wrapper logic removed

### High-level impact

- Multi-valued fields behave consistently between DB storage and CLI/template usage (`'; '` for user input/output, DB delimiter internally).
- Diff output for list fields is much more readable, showing explicit `+`/`-` item-level changes instead of generic string diffs.
- Docs and tests were updated to reflect the new multi-value behavior, including `%first` usage and `beet modify` examples.

### CODE REVIEW

File: test/ui/test_field_diff.py [hunk 1]
```diff
@@ -37,7 +39,19 @@ def diff_fmt(old, new):
...
+                {"albumtypes": ["album", "compilation"]},
+                "albumtypes",
+                dedent("""
+                    albumtypes:
+                    [text_diff_removed]  - ep[/]
+                    [text_diff_added]  + album[/]
+                    [text_diff_added]  + compilation[/]
+                """).strip(),
+                id="multi_value_changed"
+            ),
```
[the remark this answers is not in the document]
Comment (snejus) [reply test/ui/test_field_diff.py:55]: Oops!

**Comment by snejus** [comment]
@amogus07 I see you've given the PR a thumbs-up. Would you mind reviewing it as well? 😆

**Comment by amogus07** [comment]
> @amogus07 I see you've given the PR a thumbs-up. Would you mind reviewing it as well? 😆

I'm not that familiar with these parts of the codebase, unfortunately... I'm just glad this is finally being addressed

**Comment by JOJ0** [comment]
I will review and test this later today!

File: beets/dbcore/types.py [hunk 2]
```diff
@@ -288,25 +288,37 @@ class String(BaseString[str, Any]):
...
         if not string:
             return []
-        return string.split(self.delimiter)
+
+        delimiter = (
+            self.db_delimiter
+            if self.db_delimiter in string
+            else self.fmt_delimiter
+        )
+        return string.split(delimiter)
```
[the remark this answers is not in the document]
Comment (JOJ0) [reply beets/dbcore/types.py:320]: this concern is definitely neglect-able

File: beets/dbcore/types.py [hunk 3]
```diff
@@ -288,25 +288,37 @@ class String(BaseString[str, Any]):
...
     def parse(self, string: str):
         if not string:
             return []
-        return string.split(self.delimiter)
+
+        delimiter = (
+            self.db_delimiter
+            if self.db_delimiter in string
+            else self.fmt_delimiter
+        )
```
Comment (JOJ0) [inline beets/dbcore/types.py:319]: Do I get this right? With this we also support reading as multi-valued-fields data that is (incorrectly? deprecated?) stored in db as `; ` delimited? So in case of some very weird artist name for example this could be an issue? MIght be an esoteric concern though...just thinking out loud.

Comment (JOJ0) [reply beets/dbcore/types.py:319]: And yes AFAIR we had `; ` as the DB delimiter with an early implementation of `albumtypes`. I don't recall the whole history but do we still use that? Or is that long gone and I'm just not up to speed?

File: beets/ui/__init__.py [hunk 4]
```diff
@@ -571,15 +571,16 @@ def get_color_config() -> dict[ColorName, str]:
...
-    Returns colored text when color output is enabled and NO_COLOR environment
-    variable is not set, otherwise returns plain text unchanged.
-    """
+
+def colorize(color_name: ColorName, text: str) -> str:
+    """Colorize text when color output is enabled."""
     if config["ui"]["color"] and "NO_COLOR" not in os.environ:
-        color_code = get_color_config()[color_name]
-        return f"{COLOR_ESCAPE}[{color_code}m{text}{RESET_COLOR}"
+        return _colorize(color_name, text)
```
Comment (JOJ0) [inline beets/ui/__init__.py:583]: When reading these two functions in isolation it feels slightly over engineered. If I got it right, colorize is the only caller of _colorize, and splitting the color-handling-implementation from the config-handling (which is what I suppose you are trying to do) feels not too necessary.

File: beets/ui/__init__.py [hunk 5]
```diff
@@ -682,16 +663,6 @@ def _colordiff(a: Any, b: Any) -> tuple[str, str]:
...
-def colordiff(a, b):
-    """Colorize differences between two values if color is enabled.
-    (Like _colordiff but conditional.)
-    """
-    if config["ui"]["color"]:
-        return _colordiff(a, b)
-    else:
-        return str(a), str(b)
-
-
```
Comment (JOJ0) [inline beets/ui/__init__.py:694]: Hmm I see, this is where this pattern is coming from....ok...but now it's gone here so.... :-)

File: beets/ui/__init__.py [hunk 6]
```diff
@@ -1048,6 +1019,18 @@ def print_newline_layout(
...
+    removed = oldset - newset
+
+    parts = [
+        f"{field}:",
+        *(colorize("text_diff_removed", f"  - {i}") for i in sorted(removed)),
+        *(colorize("text_diff_added", f"  + {i}") for i in sorted(added)),
+    ]
+    return "\n".join(parts)
+
+
```
Comment (JOJ0) [inline beets/ui/__init__.py:1033]: This is so beautiful on my CLI ❤️ Thanks!

File: docs/reference/pathformat.rst [hunk 7]
```diff
@@ -75,11 +75,34 @@ These functions are built in to beets:
...
   by strftime_. For example, to get the year some music was added to your
   library, use ``%time{$added,%Y}``.
-- ``%first{text}``: Returns the first item, separated by ``;`` (a semicolon
-  followed by a space). You can use ``%first{text,count,skip}``, where ``count``
-  is the number of items (default 1) and ``skip`` is number to skip (default 0).
-  You can also use ``%first{text,count,skip,sep,join}`` where ``sep`` is the
-  separator, like ``;`` or ``/`` and join is the text to concatenate the items.
+- ``%first{text,count,skip,sep,join}``: Extract a subset of items from a
+  delimited string. Splits ``text`` by ``sep``, skips the first ``skip`` items,
+  then returns the next ``count`` items joined by ``join``.
```
Comment (JOJ0) [inline docs/reference/pathformat.rst:80]: I did not play around with this feature but I see we have tests for it (which you've adjusted)

File: docs/reference/pathformat.rst [hunk 8]
```diff
@@ -75,11 +75,34 @@ These functions are built in to beets:
...
+
+  Examples:
+
+  ::
+
+      %first{$genres}             →  returns the first genre
+      %first{$genres,2}           →  returns the first two genres, joined by "; "
+      %first{$genres,2,1}         →  skips the first genre, returns the next two
+      %first{$genres,2,0, , -> }  →  splits by space, joins with " -> "
+
```
Comment (JOJ0) [inline docs/reference/pathformat.rst:105]: Wow, ok this is now very thorough documentation for the feature. Thanks! Really great! 👍 I won't read the original docs a 3rd time now - I don't need to understand them anymore 🤣 🫢

File: test/ui/test_field_diff.py [hunk 9]
```diff
@@ -15,7 +17,7 @@ def configure_color(self, config, color):
     def patch_colorize(self, monkeypatch):
         """Patch to return a deterministic string format instead of ANSI codes."""
         monkeypatch.setattr(
-            "beets.ui.colorize",
+            "beets.ui._colorize",
```
Comment (JOJ0) [inline test/ui/test_field_diff.py:20]: Oh, I see, this might be the sole reason why you split it into two functions, right?

File: beets/dbcore/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JOJ0) [reply beets/dbcore/types.py:320]: Are we testing 1 and 2?

3 not important

4 not sure

Ah, edit: Lost the context here. I was referring to the Copilot concerns you had marked as resolved already. Copy pasting:

```
.... tests that verify:

1. Parsing user input with ;  delimiter works correctly
2. Parsing database values with \␀ delimiter still works
3. Edge cases like strings containing both delimiters are handled correctly
4. Empty strings return empty lists
```

File: test/test_plugins.py [hunk 10]
```diff
@@ -96,8 +96,7 @@ def test_multi_value_flex_field_type(self):
         item.add(self.lib)
 
         out = self.run_with_output("ls", "-f", "$multi_value")
-        delimiter = types.MULTI_VALUE_DSV.delimiter
-        assert out == f"one{delimiter}two{delimiter}three\n"
+        assert out == "one; two; three\n"
```
Comment (JOJ0) [inline test/test_plugins.py:99]: Should we add something like this? (but maybe not in `test_plugins.py`) Or do we cover this elsewhere?Sorry didn't look for it thoroughly.

```
def test_delimitedstring_parse():
    ds = DelimitedString(db_delimiter="\0")
    assert ds.parse("one\0two\0three") == ["one", "two", "three"]
    assert ds.parse("one; two; three") == ["one", "two", "three"]
    assert ds.parse("") == []
```

**Comment by JOJ0** [comment]
This failing test seems unrelated: https://github.com/beetbox/beets/actions/runs/22280727644/job/64450831853?pr=6387

File: beets/dbcore/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (snejus) [reply beets/dbcore/types.py:320]: So I split the conditional to measure coverage across each branch:

```py
    def parse(self, string: str):
        if not string:
            return []

        if self.db_delimiter in string:
            delimiter = self.db_delimiter
        else:
            delimiter = self.fmt_delimiter
        return string.split(delimiter)
```

And then I ran `poe test-with-coverage`.

Now, opening coverage report in `.reports/html/index.html` and navigating to see this source code, I see how many tests cover each:

```py

        if self.db_delimiter in string:     # 156 tests
            delimiter = self.db_delimiter   # 9 tests
        else:
            delimiter = self.fmt_delimiter  # 148 tests
```

This means that we have

> 1. Parsing user input with ; delimiter works correctly       # 148 tests
> 2. Parsing database values with \␀ delimiter still works     # 9 tests

Now specifically for **2.**, I can see the following tests:

```
test/test_importer.py::ImportCompilationTest::test_asis_albumartists_tag_sets_multi_albumartists
test/test_plugins.py::TestPluginRegistration::test_multi_value_flex_field_type
test/test_query.py::TestGet::test_fast_vs_slow[functools.partial(<class 'beets.dbcore.query.MatchQuery'>, 'artist', 'one')]
test/test_query.py::TestGet::test_fast_vs_slow[functools.partial(<class 'beets.dbcore.query.NoneQuery'>, 'rg_track_gain')]
test/test_query.py::TestGet::test_fast_vs_slow[functools.partial(<class 'beets.dbcore.query.NumericQuery'>, 'year', '2002')]
test/test_query.py::TestGet::test_fast_vs_slow[functools.partial(<class 'beets.dbcore.query.RegexpQuery'>, 'album', '^.a')]
test/test_query.py::TestGet::test_fast_vs_slow[functools.partial(<class 'beets.dbcore.query.StringQuery'>, 'year', '2001')]
test/ui/commands/test_update.py::UpdateTest::test_multivalued_albumtype_roundtrip
test/ui/test_ui_importer.py::ImportCompilationTest::test_asis_albumartists_tag_sets_multi_albumartists
```

File: beets/dbcore/types.py (hunk 3 above)
Comment (snejus) [reply beets/dbcore/types.py:319]: Note `albumtypes` are still stored with `; ` as the delimiter:

```py
# types.py
...
SEMICOLON_SPACE_DSV = DelimitedString("; ")

# Will set the proper null char in mediafile
MULTI_VALUE_DSV = DelimitedString("\\␀")
```

```py
# models.py
...
        "albumtypes": types.SEMICOLON_SPACE_DSV,
...
        "albumtypes": types.SEMICOLON_SPACE_DSV,
...
```

Comment (snejus) [reply beets/dbcore/types.py:319]: Essentially, this change means: _regardless_ of the delimiter we use to store a field in the database, we _expect_ the users to use `; ` as the delimiter in their input.

File: test/ui/test_field_diff.py (hunk 9 above)
Comment (snejus) [reply test/ui/test_field_diff.py:20]: You got it exactly right, my friend!

File: docs/reference/pathformat.rst (hunk 8 above)
Comment (snejus) [reply docs/reference/pathformat.rst:105]: Thanks to AI, as usual, for the documentation!

File: test/test_plugins.py (hunk 10 above)
Comment (snejus) [reply test/test_plugins.py:99]: See my comment above where I measured coverage for this!

**Comment by snejus** [comment]
Actually, I thinkw we'd also like to adjust `edit` plugin:

Currently we have:

```yaml
album: RAVE SLUTZ
album_id: 11492
albumartist: Fallen Shrine & dj Christian NXC
albumstatus: Official
albumtype: album
albumtypes: album
artist: Fallen Shrine & dj Christian NXC
artist_sort: ''
artists: Fallen Shrine; dj Christian NXC
```

While I imagine we'd instead like to have:

```yaml
album: RAVE SLUTZ
album_id: 11492
albumartist: Fallen Shrine & dj Christian NXC
albumstatus: Official
albumtype: album
albumtypes:
- album
artist: Fallen Shrine & dj Christian NXC
artist_sort: ''
artists:
- Fallen Shrine
- dj Christian NXC
```

**Review by JOJ0 — APPROVED** [review]
Let's go ahead then 😍

**Comment by JOJ0** [comment]
> Actually, I thinkw we'd also like to adjust `edit` plugin:
> 
> 
> 
> Currently we have:
> 
> 
> 
> ```yaml
> 
> album: RAVE SLUTZ
> 
> album_id: 11492
> 
> albumartist: Fallen Shrine & dj Christian NXC
> 
> albumstatus: Official
> 
> albumtype: album
> 
> albumtypes: album
> 
> artist: Fallen Shrine & dj Christian NXC
> 
> artist_sort: ''
> 
> artists: Fallen Shrine; dj Christian NXC
> 
> ```
> 
> 
> 
> While I imagine we'd instead like to have:
> 
> 
> 
> ```yaml
> 
> album: RAVE SLUTZ
> 
> album_id: 11492
> 
> albumartist: Fallen Shrine & dj Christian NXC
> 
> albumstatus: Official
> 
> albumtype: album
> 
> albumtypes:
> 
> - album
> 
> artist: Fallen Shrine & dj Christian NXC
> 
> artist_sort: ''
> 
> artists:
> 
> - Fallen Shrine
> 
> - dj Christian NXC
> 
> ```

Hmm interesting idea, since it's yaml...

Do this in a subsequent PR? For now it's good enough in my opinion. Played with edit before and didn't miss anything, but yeah would be a great and self-explanatory feature!

**Comment by snejus** [comment]
It kinds of belongs here, what do you think? The change was just two lines, so I added it here: 

[asciicast](https://asciinema.org/a/SOlAb3HdFSdkdh6r)

Note: I use `edit` quite a lot 😅

**Comment by JOJ0** [comment]
sure! edit! sorry was busy this weekend orherwise. good you included it! thanks!
